File: troTHU/input_safety.py
"""Field-aware input cleanup for human-facing CLI and monitor console paths."""

from __future__ import annotations

import contextlib
import re
import sys
from dataclasses import dataclass, field
from typing import Any, Dict, List, Mapping


SENSITIVE_FIELD_TYPES = {"password", "token", "secret"}
QR_FIELD_TYPES = {"qr_payload", "qr"}
COLLAPSE_SPACE_RE = re.compile(r"\s+")
TIME_RANGE_RE = re.compile(r"\b\d{1,2}[:：]\d{2}\b")


@dataclass(frozen=True)
class InputSanitizationResult:
    value: str
    changed: bool = False
    warnings: tuple[str, ...] = field(default_factory=tuple)
    valid: bool = True
    reason: str = ""

    def to_dict(self) -> Dict[str, Any]:
        return {
            "value": "[redacted]" if self.reason == "sensitive" else self.value,
            "changed": self.changed,
            "warnings": list(self.warnings),
            "valid": self.valid,
            "reason": self.reason,
        }


def _warning(field: str, message: str) -> str:
    return "{}: {}".format(field, message)


def _collapse_spaces(value: str) -> str:
    return COLLAPSE_SPACE_RE.sub(" ", value).strip()


def _normalize_field_type(field_type: str) -> str:
    return str(field_type or "text").strip().lower().replace("-", "_")


def sanitize_input_field(value: Any, *, field_type: str = "text", field_name: str = "") -> InputSanitizationResult:
    kind = _normalize_field_type(field_type)
    label = field_name or kind
    original = "" if value is None else str(value)
    warnings: List[str] = []

    if kind in QR_FIELD_TYPES:
        cleaned = original.strip()
        if cleaned != original:
            warnings.append(_warning(label, "已移除前後空白；QR 內容內部未被修改。"))
        return InputSanitizationResult(cleaned, cleaned != original, tuple(warnings), bool(cleaned), "empty" if not cleaned else "")

    if kind in SENSITIVE_FIELD_TYPES:
        cleaned = original.strip()
        if cleaned != original:
            warnings.append(_warning(label, "已移除前後空白；內容已遮蔽，不會記錄原值。"))
        return InputSanitizationResult(cleaned, cleaned != original, tuple(warnings), bool(cleaned), "sensitive" if cleaned else "empty")

    if kind in {"student_id", "profile", "profile_name", "provider", "env", "env_var", "channel_id", "host", "port"}:
        cleaned = _collapse_spaces(original)
    elif kind in {"time_range", "schedule"}:
        text = original.replace("：", ":")
        matches = TIME_RANGE_RE.findall(text)
        cleaned = " - ".join(item.replace("：", ":") for item in matches[:2]) if len(matches) >= 2 else _collapse_spaces(text)
    else:
        cleaned = _collapse_spaces(original)

    if cleaned != original:
        warnings.append(_warning(label, "已自動修正多餘空白。"))

    valid = True
    reason = ""
    if kind in {"profile", "profile_name"} and not cleaned:
        valid = False
        reason = "empty_profile"
        warnings.append(_warning(label, "profile 名稱不可為空。"))
    if kind == "port":
        try:
            port = int(cleaned)
            valid = 1 <= port <= 65535
        except (TypeError, ValueError):
            valid = False
        if not valid:
            reason = "invalid_port"
            warnings.append(_warning(label, "port 必須是 1 到 65535 的整數。"))
    if kind == "provider" and cleaned:
        cleaned = cleaned.lower()

    return InputSanitizationResult(cleaned, cleaned != original, tuple(warnings), valid, reason)


def sanitize_mapping_fields(value: Mapping[str, Any], field_types: Mapping[str, str]) -> tuple[Dict[str, Any], List[str]]:
    sanitized: Dict[str, Any] = dict(value)
    warnings: List[str] = []
    for key, field_type in field_types.items():
        if key not in sanitized:
            continue
        result = sanitize_input_field(sanitized.get(key), field_type=field_type, field_name=key)
        sanitized[key] = result.value
        warnings.extend(result.warnings)
    return sanitized, warnings


def sanitize_config_values(config: Dict[str, Any]) -> List[str]:
    """Mutate common config string fields into safer human-entered values."""
    warnings: List[str] = []
    if not isinstance(config, dict):
        return warnings

    account = config.get("account")
    if isinstance(account, dict):
        sanitized, field_warnings = sanitize_mapping_fields(
            account,
            {"user": "student_id", "passwd": "password"},
        )
        account.update(sanitized)
        warnings.extend(field_warnings)

    accounts = config.get("accounts")
    profiles = accounts.get("profiles") if isinstance(accounts, dict) else {}
    if isinstance(profiles, dict):
        for name, profile in list(profiles.items()):
            safe_name = sanitize_input_field(name, field_type="profile", field_name="accounts.profile").value
            if safe_name and safe_name != name:
                profiles[safe_name] = profiles.pop(name)
                warnings.append("accounts.profile: 已修正 profile 名稱空白。")
            if isinstance(profile, dict):
                sanitized, field_warnings = sanitize_mapping_fields(
                    profile,
                    {"user": "student_id", "passwd": "password", "label": "text"},
                )
                profile.update(sanitized)
                warnings.extend(field_warnings)
        if isinstance(accounts.get("current"), str):
            result = sanitize_input_field(accounts.get("current"), field_type="profile", field_name="accounts.current")
            accounts["current"] = result.value or "default"
            warnings.extend(result.warnings)

    provider = config.get("provider")
    if isinstance(provider, dict):
        for key in ("current", "requested"):
            if key in provider:
                result = sanitize_input_field(provider.get(key), field_type="provider", field_name="provider.{}".format(key))
                provider[key] = result.value
                warnings.extend(result.warnings)

    local_ui = config.get("local_ui")
    if isinstance(local_ui, dict):
        sanitized, field_warnings = sanitize_mapping_fields(local_ui, {"host": "host", "port": "port"})
        local_ui.update(sanitized)
        warnings.extend(field_warnings)

    integrations = config.get("integrations")
    if isinstance(integrations, dict):
        for adapter, keys in {
            "discord": {"token_env": "env_var", "channel_env": "env_var", "public_key_env": "env_var", "application_id_env": "env_var", "guild_id_env": "env_var"},
            "line": {"token_env": "env_var", "secret_env": "env_var"},
            "telegram": {"token_env": "env_var", "chat_env": "env_var"},
        }.items():
            item = integrations.get(adapter)
            if isinstance(item, dict):
                sanitized, field_warnings = sanitize_mapping_fields(item, keys)
                item.update(sanitized)
                warnings.extend(field_warnings)

    return warnings


def contains_sensitive_text(value: Any) -> bool:
    text = str(value or "").lower()
    return any(part in text for part in ("password", "passwd", "token", "secret", "cookie", "session", "payload"))


def _fallback_password_input(prompt: str) -> str:
    try:
        return input(prompt).strip()
    except (EOFError, KeyboardInterrupt):
        return ""


def _optional_status_line_pause():
    try:
        import troTHU.runtime_context as ctx  # type: ignore
    except Exception:
        try:
            import runtime_context as ctx  # type: ignore
        except Exception:
            return contextlib.nullcontext()
    try:
        return ctx.pause_status_line()
    except Exception:
        return contextlib.nullcontext()


def masked_password_input(prompt: str = "輸入密碼 > ") -> str:
    with _optional_status_line_pause():
        return _masked_password_input(prompt)


def _masked_password_input(prompt: str = "輸入密碼 > ") -> str:
    """Read a password with a best-effort local mask without using getpass."""
    prompt_text = str(prompt or "輸入密碼 > ")
    if sys.platform.startswith("win"):
        try:
            import msvcrt  # type: ignore

            sys.stdout.write(prompt_text)
            sys.stdout.flush()
            chars: List[str] = []
            while True:
                ch = msvcrt.getwch()
                if ch in ("\r", "\n"):
                    sys.stdout.write("\n")
                    sys.stdout.flush()
                    return "".join(chars).strip()
                if ch == "\x03":
                    raise KeyboardInterrupt
                if ch in ("\b", "\x7f"):
                    if chars:
                        chars.pop()
                        sys.stdout.write("\b \b")
                        sys.stdout.flush()
                    continue
                if ch in ("\x00", "\xe0"):
                    msvcrt.getwch()
                    continue
                if ch and ch.isprintable():
                    chars.append(ch)
                    sys.stdout.write("*")
                    sys.stdout.flush()
        except (EOFError, KeyboardInterrupt):
            sys.stdout.write("\n")
            sys.stdout.flush()
            return ""
        except Exception:
            return _fallback_password_input(prompt_text)

    try:
        import termios
        import tty

        stdin = sys.stdin
        fd = stdin.fileno()
        old_settings = termios.tcgetattr(fd)
        sys.stdout.write(prompt_text)
        sys.stdout.flush()
        chars = []
        try:
            tty.setraw(fd)
            while True:
                ch = stdin.read(1)
                if ch in ("\r", "\n"):
                    sys.stdout.write("\n")
                    sys.stdout.flush()
                    return "".join(chars).strip()
                if ch == "\x03":
                    raise KeyboardInterrupt
                if ch in ("\b", "\x7f"):
                    if chars:
                        chars.pop()
                        sys.stdout.write("\b \b")
                        sys.stdout.flush()
                    continue
                if ch and ch.isprintable():
                    chars.append(ch)
                    sys.stdout.write("*")
                    sys.stdout.flush()
        finally:
            termios.tcsetattr(fd, termios.TCSADRAIN, old_settings)
    except (EOFError, KeyboardInterrupt):
        sys.stdout.write("\n")
        sys.stdout.flush()
        return ""
    except Exception:
        return _fallback_password_input(prompt_text)

File: src/security/input-safety.js
'use strict';

/**
 * Field-aware input cleanup for human-facing CLI and monitor console paths.
 */

const SENSITIVE_FIELD_TYPES = new Set(['password', 'token', 'secret']);
const QR_FIELD_TYPES = new Set(['qr_payload', 'qr']);
const COLLAPSE_SPACE_RE = /\s+/g;
const TIME_RANGE_RE = /\b\d{1,2}[:：]\d{2}\b/g;

// ---------------------------------------------------------------------------
// InputSanitizationResult
// ---------------------------------------------------------------------------

class InputSanitizationResult {
  /**
   * @param {string} value
   * @param {boolean} [changed=false]
   * @param {string[]} [warnings=[]]
   * @param {boolean} [valid=true]
   * @param {string} [reason='']
   */
  constructor(value, changed = false, warnings = [], valid = true, reason = '') {
    this.value = value;
    this.changed = changed;
    this.warnings = Object.freeze([...warnings]);
    this.valid = valid;
    this.reason = reason;
    Object.freeze(this);
  }

  toDict() {
    return {
      value: this.reason === 'sensitive' ? '[redacted]' : this.value,
      changed: this.changed,
      warnings: [...this.warnings],
      valid: this.valid,
      reason: this.reason,
    };
  }
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

function _warning(field, message) {
  return `${field}: ${message}`;
}

function _collapseSpaces(value) {
  return value.replace(COLLAPSE_SPACE_RE, ' ').trim();
}

function _normalizeFieldType(fieldType) {
  return String(fieldType || 'text').trim().toLowerCase().replace(/-/g, '_');
}

// ---------------------------------------------------------------------------
// Core Sanitization
// ---------------------------------------------------------------------------

/**
 * @param {*} value
 * @param {Object} [opts]
 * @param {string} [opts.fieldType='text']
 * @param {string} [opts.fieldName='']
 * @returns {InputSanitizationResult}
 */
function sanitizeInputField(value, { fieldType = 'text', fieldName = '' } = {}) {
  const kind = _normalizeFieldType(fieldType);
  const label = fieldName || kind;
  const original = value == null ? '' : String(value);
  const warnings = [];

  if (QR_FIELD_TYPES.has(kind)) {
    const cleaned = original.trim();
    if (cleaned !== original) {
      warnings.push(_warning(label, '已移除前後空白；QR 內容內部未被修改。'));
    }
    return new InputSanitizationResult(cleaned, cleaned !== original, warnings, Boolean(cleaned), cleaned ? '' : 'empty');
  }

  if (SENSITIVE_FIELD_TYPES.has(kind)) {
    const cleaned = original.trim();
    if (cleaned !== original) {
      warnings.push(_warning(label, '已移除前後空白；內容已遮蔽，不會記錄原值。'));
    }
    return new InputSanitizationResult(cleaned, cleaned !== original, warnings, Boolean(cleaned), cleaned ? 'sensitive' : 'empty');
  }

  let cleaned;
  if (['student_id', 'profile', 'profile_name', 'provider', 'env', 'env_var', 'channel_id', 'host', 'port'].includes(kind)) {
    cleaned = _collapseSpaces(original);
  } else if (['time_range', 'schedule'].includes(kind)) {
    const text = original.replace(/：/g, ':');
    const matches = text.match(TIME_RANGE_RE);
    cleaned = (matches && matches.length >= 2)
      ? matches.slice(0, 2).map(m => m.replace(/：/g, ':')).join(' - ')
      : _collapseSpaces(text);
  } else {
    cleaned = _collapseSpaces(original);
  }

  if (cleaned !== original) {
    warnings.push(_warning(label, '已自動修正多餘空白。'));
  }

  let valid = true;
  let reason = '';

  if (['profile', 'profile_name'].includes(kind) && !cleaned) {
    valid = false;
    reason = 'empty_profile';
    warnings.push(_warning(label, 'profile 名稱不可為空。'));
  }

  if (kind === 'port') {
    try {
      const port = parseInt(cleaned, 10);
      valid = !Number.isNaN(port) && port >= 1 && port <= 65535;
    } catch {
      valid = false;
    }
    if (!valid) {
      reason = 'invalid_port';
      warnings.push(_warning(label, 'port 必須是 1 到 65535 的整數。'));
    }
  }

  if (kind === 'provider' && cleaned) {
    cleaned = cleaned.toLowerCase();
  }

  return new InputSanitizationResult(cleaned, cleaned !== original, warnings, valid, reason);
}

/**
 * @param {Object} value
 * @param {Object<string, string>} fieldTypes
 * @returns {{ sanitized: Object, warnings: string[] }}
 */
function sanitizeMappingFields(value, fieldTypes) {
  const sanitized = { ...value };
  const warnings = [];
  for (const [key, fieldType] of Object.entries(fieldTypes)) {
    if (!(key in sanitized)) continue;
    const result = sanitizeInputField(sanitized[key], { fieldType, fieldName: key });
    sanitized[key] = result.value;
    warnings.push(...result.warnings);
  }
  return { sanitized, warnings };
}

/**
 * Mutate common config string fields into safer human-entered values.
 * @param {Object} config
 * @returns {string[]}
 */
function sanitizeConfigValues(config) {
  const warnings = [];
  if (!config || typeof config !== 'object') return warnings;

  // account
  const account = config.account;
  if (account && typeof account === 'object') {
    const { sanitized, warnings: w } = sanitizeMappingFields(account, { user: 'student_id', passwd: 'password' });
    Object.assign(account, sanitized);
    warnings.push(...w);
  }

  // accounts.profiles
  const accounts = config.accounts;
  const profiles = (accounts && typeof accounts === 'object') ? accounts.profiles : null;
  if (profiles && typeof profiles === 'object') {
    for (const [name, profile] of Object.entries(profiles)) {
      const safeName = sanitizeInputField(name, { fieldType: 'profile', fieldName: 'accounts.profile' }).value;
      if (safeName && safeName !== name) {
        profiles[safeName] = profiles[name];
        delete profiles[name];
        warnings.push('accounts.profile: 已修正 profile 名稱空白。');
      }
      if (profile && typeof profile === 'object') {
        const { sanitized, warnings: w } = sanitizeMappingFields(profile, { user: 'student_id', passwd: 'password', label: 'text' });
        Object.assign(profile, sanitized);
        warnings.push(...w);
      }
    }
    if (typeof accounts.current === 'string') {
      const result = sanitizeInputField(accounts.current, { fieldType: 'profile', fieldName: 'accounts.current' });
      accounts.current = result.value || 'default';
      warnings.push(...result.warnings);
    }
  }

  // provider
  const provider = config.provider;
  if (provider && typeof provider === 'object') {
    for (const key of ['current', 'requested']) {
      if (key in provider) {
        const result = sanitizeInputField(provider[key], { fieldType: 'provider', fieldName: `provider.${key}` });
        provider[key] = result.value;
        warnings.push(...result.warnings);
      }
    }
  }

  // local_ui
  const localUi = config.local_ui;
  if (localUi && typeof localUi === 'object') {
    const { sanitized, warnings: w } = sanitizeMappingFields(localUi, { host: 'host', port: 'port' });
    Object.assign(localUi, sanitized);
    warnings.push(...w);
  }

  // integrations
  const integrations = config.integrations;
  if (integrations && typeof integrations === 'object') {
    const adapterFieldTypes = {
      discord: { token_env: 'env_var', channel_env: 'env_var', public_key_env: 'env_var', application_id_env: 'env_var', guild_id_env: 'env_var' },
      line: { token_env: 'env_var', secret_env: 'env_var' },
      telegram: { token_env: 'env_var', chat_env: 'env_var' },
    };
    for (const [adapter, keys] of Object.entries(adapterFieldTypes)) {
      const item = integrations[adapter];
      if (item && typeof item === 'object') {
        const { sanitized, warnings: w } = sanitizeMappingFields(item, keys);
        Object.assign(item, sanitized);
        warnings.push(...w);
      }
    }
  }

  return warnings;
}

/**
 * @param {*} value
 * @returns {boolean}
 */
function containsSensitiveText(value) {
  const text = String(value || '').toLowerCase();
  return ['password', 'passwd', 'token', 'secret', 'cookie', 'session', 'payload'].some(p => text.includes(p));
}

/**
 * Masked password input for CLI.
 * @param {string} [prompt='輸入密碼 > ']
 * @returns {Promise<string>}
 */
async function maskedPasswordInput(prompt = '輸入密碼 > ') {
  const readline = require('node:readline');
  const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout,
    terminal: true,
  });

  return new Promise((resolve) => {
    // Mute output for password masking
    const origWrite = process.stdout.write.bind(process.stdout);
    let muted = false;
    process.stdout.write = function (chunk, encoding, callback) {
      if (muted) {
        // Write asterisk for each character (skip control chars)
        if (typeof chunk === 'string' && chunk.length === 1 && chunk.charCodeAt(0) >= 32) {
          return origWrite('*', encoding, callback);
        }
        return origWrite(chunk, encoding, callback);
      }
      return origWrite(chunk, encoding, callback);
    };

    rl.question(prompt, (answer) => {
      muted = false;
      process.stdout.write = origWrite;
      rl.close();
      resolve(String(answer || '').trim());
    });
    muted = true;
  });
}

module.exports = {
  InputSanitizationResult,
  sanitizeInputField,
  sanitizeMappingFields,
  sanitizeConfigValues,
  containsSensitiveText,
  maskedPasswordInput,
  SENSITIVE_FIELD_TYPES,
  QR_FIELD_TYPES,
};
